Add Progress SPEC (#24418)
microsoft/fluentui: packages/react-components/react-progress/Spec.md

| Lightning Design System | Progress Bar       | [ProgressBar](https://www.lightningdesignsystem.com/components/progress-bar/)                                | Has a vertical bar, only determinate, and can specify progress step                               |
| Semantic UI             | Progress           | [Progress](https://semantic-ui.com/modules/progress.html#indicating)                                         | Allows for success and error states, default rounded edges. No indeterminate form                 |

### Comparison of v8 and v0

The existing components are:

- v8
  - `ProgressIndicator`
- v0

## Sample Code

Basic example:

```jsx
import { Progress } from '@fluentui/react-progress';

function App() {
  return <Progress thickness="large" label="Loading" />;
}
```

## Variants

- Indeterminate Progress
  - The default Progress that animates indefinitely
- Determinate Progress
  - The determinate form of the Progress component that incrementally loads from 0% to 100%

### Shape

The Progress is represented as a rounded rectangular area with an inner animated bar that either travels across the area indefinitely or animates up till a specified point

## API

> ### Slots

Reviewer: We might want to use `Field` to add support for the label and description. It also looks like the visual spec includes support for error and success states, which the Field component already supports.

That would mean:
* The `Progress` control only has two slots:
  * `root` (which would be the track)
  * `bar` (child of root)
* There's a new `ProgresField` control that wraps `Progress` and adds the features of Field: label, helperText, validationState, etc.

That would follow the pattern of our other form controls, which (mostly) don't include labels built-in. And it would also benefit from sharing all of the features of Field.

Reviewer 2: I agree with this but I think we discussed maybe making those changes to both spec and implementation in a separate PR.

- `root` - The root element of the Progress. The html element is a `div`
- `bar` - The div element that gets animated into a Progress bar. The html element is `div`
- `track` - The div element that functions as the track for the Progress bar. The html element is `div`
- `label` - The text shown above the Progress. The html element is a `span`
- `description` - The text shown below the Progress. The html element is a `span`

### Props

> See API at [Progress.types.tsx](./src/components/Progress/Progress.types.ts).

Reviewer: I wonder if we should make the props align with the [HTML `<progress>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress):
* Use `value` instead of `percentComplete`
* Add a `max` prop that defaults to 1.0
* Remove the `indeterminate` prop, and just use the `value` prop: the progress bar is indeterminate if value is undefined.

## Background

> A `Progress` is a component that is used to display the current progress of an operation flow.

Reviewer 2: _nit:_ Wording

```suggestion
The `Progress` component is used to display the current progress of an operation flow.
```

## Prior Art

> ### Open UI

Reviewer 3: I don't see a reference there to [OpenUI](https://open-ui.org/), is the title correct?

Reviewer 2: @tomi-msft you can add a link to Open UI if there is a page about `Progress` there. If a page for progress does not exist feel free to rename this section to something like "Comparison between other libraries".

| Library    | Component Name | Spec Link                                                          | Notes                                                                   |
| ---------- | -------------- | ------------------------------------------------------------------ | ----------------------------------------------------------------------- |
| Ant Design | Progress       | [Progress](https://ant.design/components/progress/)                | Specifies the type, a combined progress component                       |

> | Bootstrap  | Progress       | [Progress](https://getbootstrap.com/docs/4.3/components/progress/) | Allows for multiple different animation styles on the same Progress bar |

Reviewer 3: I am surprised with such small list of prior art as this component is quite popular:
- https://react.semantic-ui.com/modules/progress/
- https://react-spectrum.adobe.com/react-spectrum/ProgressBar.html
- https://vuetifyjs.com/en/components/progress-linear/
- https://mui.com/material-ui/react-progress/

Did you consider other libraries in a research?

### Comparison of v8 and v0

The existing components are:

- v8
  - `ProgressIndicator`
- v0

## Sample Code

> Basic examples:

Reviewer 3: ```suggestion
Basic example:
```

_nit_ There is a single example now 🐱

> ```jsx=

Reviewer 3: ````suggestion
```jsx
````

It seems that the snippet does not have proper formatting:

![image](https://user-images.githubusercontent.com/14183168/186391940-2847e92c-27fc-40f3-9f7f-ab16008b7e9f.png)

import { Progress } from "@fluentui/react-progress";

function App() {
    return (
        <Progress barHeight="medium" label="Loading" />
    )
}

> 
> ```

Reviewer 3: ````suggestion
```
````

## Variants

- Indeterminate Progress
  - The default Progress that animates indefinitely
- Determinate Progress
  - The determinate form of the Progress component that incrementally loads from 0 to 100

### Shape

The Progress is represented as a rectangular area with an inner animated bar that either travels across the area indefinitely or animates up till a specified point

## API

> ### Slots

Reviewer 3: It would be great to include there an image with a component's anatomy so it would be clearer which slots correspond to what area.

- `root` - The root element of the Progress. The html element is a `div`

> - `indicator` - The div element that gets animated into a Progress. The html element is `span`

Reviewer 3: I know that we call this part `indicator` in few other places, but few other libraries call this part `bar`. _Curiosity_: what is a motivation to call it `indicator`? 🐱 Taking into account that there is a prop called `barThickness` in #24023.

- `label` - The text shown above the Progress. This uses the `Label` control
- `description` - The text shown below the Progress. This uses the `Label` control

### Props

> See API at [Progress.types.tsx](./src/components/Progress/Progress.types.ts).

Reviewer 3: FYI: it's currently empty:

![image](https://user-images.githubusercontent.com/14183168/186392360-2e50fdc0-8cce-42c7-b77b-3933f56e169e.png)

It would be great to include currently implemented props from #24023 to the spec.

## Structure

```html
<div class="fui-Progress">
  <!-- Label for Progress -->
  <label className="fui-Progress__label">Loading...</label>
  <span className="fui-Progress__indicator" role="progressbar">
    <!-- div that receives the animation. Classnames are used for animation -->
    <div className="fui-Progress__Container">
      <div className="fui-Progress__Track" />
      <div className="fui-Progress__Tail" />
    </div>
  </span>
  <!-- Label for Progress description -->
  <label className="fui-Progress__description">Loading Text</label>
</div>

> ```

Reviewer 3: From markup I see that we will not allow to pass `children`. Is it correct? If so, I would mention this somewhere in the spec.

## Migration

See [MIGRATION.md](./MIGRATION.md).

## Behaviors

### States

- **Display** - The Progress will use the following priority:

>   - Adding `appearance={inverted}` will alter the way that the Progress is displayed.

Reviewer 2: This says that `appearance={inverted}` will alter the way `Progress` is displayed but doesn't say how. Can we add that?

>   - Adding `determinate` and specifying the `percentComplete` from `0` to `1` will alter the Progress from indeterminate to determinate

Reviewer 2: _nit:_

```suggestion
  - Adding `determinate` and specifying the `percentComplete` from `0` to `1` will alter the Progress from indeterminate to determinate.
```

>   - The Progress also has `rtl` support that will animate the progress bar from right to left

Reviewer 2: _nit:_ wording

```suggestion
  - The component also has `rtl` support and will animate the progress bar from right to left if set.
```
